module/zfs/zfs_ioctl.c

```diff
@@ -6579,6 +6747,83 @@ zfs_ioctl_init(void)
 	    zfs_secpolicy_config, NO_NAME, B_FALSE, POOL_CHECK_NONE);
 }
 
+/*
+ * Verify that for non-legacy ioctls the input nvlist
+ * pairs match against the expected input.
+ *
+ * Possible errors are:
+ * ZFS_ERR_IOC_ARG_UNAVAIL	An unrecognized nvpair was encountered
+ * ZFS_ERR_IOC_ARG_REQUIRED	A required nvpair is missing
+ * ZFS_ERR_IOC_ARG_BADTYPE	Invalid type for nvpair
+ */
+static int
+zfs_check_input_nvpairs(nvlist_t *innvl, const zfs_ioc_vec_t *vec)
+{
+	const zfs_ioc_key_t *nvl_keys = vec->zvec_nvl_keys;
+	uint64_t required_found_pairs = 0;
+
+	/*
+	 * examine each input pair
+	 */
+	for (nvpair_t *pair = nvlist_next_nvpair(innvl, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(innvl, pair)) {
+		char *name = nvpair_name(pair);
+		data_type_t type = nvpair_type(pair);
+		boolean_t identified = B_FALSE;
+
+		/*
+		 * check pair against the documented names and type
+		 */
+		for (int k = 0; k < vec->zvec_nvl_key_count; k++) {
+			/* if not a wild card name, check for an exact match */
+			if ((nvl_keys[k].zkey_flags & ZK_WILDCARDLIST) == 0 &&
+			    strcmp(nvl_keys[k].zkey_name, name) != 0)
+				continue;
+
+			identified = B_TRUE;
+
+			if (nvl_keys[k].zkey_type != DATA_TYPE_UNKNOWN &&
+			    nvl_keys[k].zkey_type != type) {
+				zfs_dbgmsg("ioc %s: pair '%s' has unexpected "
+				    "type %d (expecting %d)", vec->zvec_name,
+				    name, (int)type,
+				    (int)nvl_keys[k].zkey_type);
+				return (SET_ERROR(ZFS_ERR_IOC_ARG_BADTYPE));
+			}
+
+			if (nvl_keys[k].zkey_flags & ZK_OPTIONAL)
+				continue;
+
+			/* encountered a required key */
+			required_found_pairs |= 1 << k;
+			break;
+		}
+
+		/* allow an 'optional' key, everything else is invalid */
+		if (!identified && strcmp(name, "optional") != 0) {
+			zfs_dbgmsg("ioc %s: unidentified input pair '%s'",
+			    vec->zvec_name, name);
```

Comment:
I'm not sure how I feel about generating dbgmsg's for these failures... this could be routine depending on the use case.